Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean

```diff
@@ -0,0 +1,298 @@
+  (W.nonsingular_iff ![0, 1, 0]).mpr ⟨W.equation_zero 1, by simp⟩
+
+lemma nonsingular_zero' [NoZeroDivisors R] {Y : R} (hy : Y ≠ 0) : W.nonsingular ![0, Y, 0] :=
+  (W.nonsingular_iff ![0, Y, 0]).mpr ⟨W.equation_zero Y, by simpa⟩
+
+lemma nonsingular_some (X Y : R) : W.nonsingular ![X, Y, 1] ↔ W.toAffine.nonsingular X Y := by
+  rw [nonsingular_iff]
+  matrix_simp
+  simp only [W.toAffine.nonsingular_iff, equation_some, and_congr_right_iff,
+    W.toAffine.equation_iff, ← not_and_or, not_iff_not, one_pow, mul_one, Iff.comm, iff_self_and]
+  intro h hX hY
+  linear_combination (norm := ring1) 3 * h - X * hX - Y * hY
+
+lemma nonsingular_smul_iff (P : Fin 3 → R) (u : Rˣ) : W.nonsingular (u • P) ↔ W.nonsingular P :=
+  have (u : Rˣ) {P : Fin 3 → R} (h : W.nonsingular <| u • P) : W.nonsingular P := by
+    rcases (W.nonsingular_iff _).mp h with ⟨h, h'⟩
+    refine (W.nonsingular_iff P).mpr ⟨(W.equation_smul_iff P u).mp h, ?_⟩
+    contrapose! h'
+    simp only [smul_fin3_ext]
+    exact ⟨by linear_combination (norm := ring1) (u : R) ^ 2 * h'.left,
+      by linear_combination (norm := ring1) (u : R) ^ 2 * h'.right.left,
+      by linear_combination (norm := ring1) (u : R) ^ 2 * h'.right.right⟩
+  ⟨this u, fun h => this u⁻¹ <| by rwa [inv_smul_smul]⟩
+
+lemma nonsingular_of_equiv {P Q : Fin 3 → R} (h : P ≈ Q) : W.nonsingular P ↔ W.nonsingular Q := by
+  rcases h with ⟨u, rfl⟩
+  exact W.nonsingular_smul_iff Q u
+
+/-- The proposition that a point class on `W` is nonsingular. If `P` is a point representative,
+then `W.nonsingular_lift ⟦P⟧` is definitionally equivalent to `W.nonsingular P`. -/
+@[pp_dot]
+def nonsingular_lift (P : PointClass R) : Prop :=
+  P.lift W.nonsingular fun _ _ => propext ∘ W.nonsingular_of_equiv
+
+@[simp]
+lemma nonsingular_lift_eq (P : Fin 3 → R) : W.nonsingular_lift ⟦P⟧ = W.nonsingular P :=
+  rfl
+
+lemma nonsingular_lift_zero [Nontrivial R] : W.nonsingular_lift ⟦![0, 1, 0]⟧ :=
+  W.nonsingular_zero
+
+lemma nonsingular_lift_zero' [NoZeroDivisors R] {Y : R} (hy : Y ≠ 0) :
+    W.nonsingular_lift ⟦![0, Y, 0]⟧ :=
+  W.nonsingular_zero' hy
+
+lemma nonsingular_lift_some (X Y : R) :
+    W.nonsingular_lift ⟦![X, Y, 1]⟧ ↔ W.toAffine.nonsingular X Y :=
+  W.nonsingular_some X Y
+
+variable {F : Type u} [Field F] {W : Projective F}
+
+lemma equiv_of_Zeq0 {P Q : Fin 3 → F} (hP : W.nonsingular P) (hQ : W.nonsingular Q) (hPz : P z = 0)
+    (hQz : Q z = 0) : P ≈ Q := by
+  rw [fin3_def P, hPz] at hP ⊢
+  rw [fin3_def Q, hQz] at hQ ⊢
+  simp [nonsingular_iff, equation_iff] at hP hQ
+  simp [pow_eq_zero hP.left.symm, pow_eq_zero hQ.left.symm] at *
+  exact ⟨Units.mk0 (P y / Q y) <| div_ne_zero hP hQ, by simp [div_mul_cancel _ hQ]⟩
+
+lemma equiv_zero_of_Zeq0 {P : Fin 3 → F} (h : W.nonsingular P) (hPz : P z = 0) : P ≈ ![0, 1, 0] :=
+  equiv_of_Zeq0 h W.nonsingular_zero hPz rfl
+
+lemma equiv_some_of_Zne0 {P : Fin 3 → F} (hPz : P z ≠ 0) : P ≈ ![P x / P z, P y / P z, 1] :=
```

Comment:
Again, maybe better to require `IsUnit (P z)` and use the inverse; then one can weaken the assumptions on `F`.
Of course, for more general rings, it is possible that the last coordinate is neither a unit nor zero, so not all cases are covered, but this will be the interesting case for example when implementing the ECM factorization method. So I think it makes sense to keep that in mind.